# C_terminal_domain_analyzer

<p style="text-align: justify;">Bacterial tyrosinases feature a type III copper center in their active domain and 
are often expressed in a latent, enzymatically inactive state. 
This latency is caused by a specialized domain which is located at the C-terminal
end of the enzyme and is, therefore, called "C-terminal domain".
While all characterized tyrosinases from plants, fungi, and metazoa feature a 
C-terminal domain, bacterial tyrosinases can be expressed without it.

<p style="text-align: justify;">This repository contains code to perform a comprehensive analysis of the C-terminal 
domain in bacterial tyrosinases.

An input dataframe is provided containing the UniProt IDs of enzymes to be investigated as well as their amino acid sequence, taxonomic lineage, host organism, and enzyme name (all of which can be downloaded from the UniProt databank). The following steps are performed by running the code:
- First, entries are comprehensively filtered to remove enzymes incorrectly downloaded as tyrosinases from the UniProt databank.
- Next, 3D structures of tyrosinase enzymes retained after filtering are automatically downloaded from AlphaFold. 
- The presence of a type III copper center, which is characteristic of tyrosinases and is composed of six conserved 
histidines coordinating two copper ions is computationally checked by accessing the .pdb structure files downloaded from AlphFold.
- The DSSP module integrated into biopython is used to compute secondary structure elements (beta-strands and alpha-helices) by accessing the .pdb structure files downloaded from AlphFold.
- Machine learning is employed to classify enzymes as featuring or lacking a C-terminal domain based on the presence or absence of secondary structure elements in the .pdb structure files. 
- In-depth analysis of identified C-terminal domains is performed, including calculating their molecular weight, the number of beta-sheets, the presence of a twin-arginine 
translocation motif, the presence of amino acids responsible for copper incorporation, and the identification of the so-called place-holder residue, which 
is critically involved in conferring the latency.
- An additional module for the investigation of caddie proteins, which are separate proteins adopting the role of a C-terminal domain in tyrosinases from *Streptomyces* species, is provided.

An .xlsx file downloaded from the UniProt databank containing all enzymes to be investigated is provided as input. 

A master file performing the full-length analysis will be provided following the
publication of results generated by this code.</p>

<img src="https://github.com/FelixP91/C_terminal_domain_analyzer/blob/master/C_terminal_domain_Figure.png?raw=true" width="600" align=center>
<p style="text-align: justify;">left: tyrosinase (cyan) lacking a C-terminal domain. Substrates (yellow, mono- or o-diphenols) can readily access the active center, where they are converted into products (o-quinones).
right: tyrosinase featuring a C-terminal domain (red). The active domain of the tyrosinase enzyme (cyan), which features the catalytic center, is in a latent state as the C-terminal domain sterically prevents substrates from accessing the active center. The two copper atoms present in the active center of tyrosinases can be partially seen as brown spheres.

## Requirements
- general: os, re, pandas, numpy, bio, datetime, time, itertools, math 
- downloading pdb files from AlphaFold: selenium
- machine learning: sklearn
- DSSP: to run the code DSSP has to be installed. It can be downloaded and placed in the working directory from:
  https://swift.cmbi.umcn.nl/gv/dssp/
  or it can be downloaded and installed using:
    
      conda install -c salilab dssp


